tirocks/src/iterator.rs

```diff
@@ -0,0 +1,272 @@
+    /// an entry that comes at or past target.
+    /// All Seek*() methods clear any error status() that the iterator had prior to
+    /// the call; after the seek, status() indicates only the error (if any) that
+    /// happened during the seek, not any past errors.
+    #[inline]
+    pub fn seek(&mut self, target: &[u8]) {
+        unsafe { tirocks_sys::crocksdb_iter_seek(self.ptr, r(target)) }
+    }
+
+    /// Position at the last key in the source that at or before target.
+    /// The iterator is Valid() after this call iff the source contains
+    /// an entry that comes at or before target.
+    #[inline]
+    pub fn seek_for_prev(&mut self, target: &[u8]) {
+        unsafe { tirocks_sys::crocksdb_iter_seek_for_prev(self.ptr, r(target)) }
+    }
+
+    /// Moves to the next entry in the source.  After this call, Valid() is
+    /// true iff the iterator was not positioned at the last entry in the source.
+    /// REQUIRES: Valid()
+    #[inline]
+    pub fn next(&mut self) {
+        unsafe { tirocks_sys::crocksdb_iter_next(self.ptr) }
+    }
+
+    /// Moves to the previous entry in the source.  After this call, Valid() is
+    /// true iff the iterator was not positioned at the first entry in source.
+    /// REQUIRES: Valid()
+    #[inline]
+    pub fn prev(&mut self) {
+        unsafe { tirocks_sys::crocksdb_iter_prev(self.ptr) }
+    }
+
+    /// Return the key for the current entry.  The underlying storage for
+    /// the returned slice is valid only until the next modification of
+    /// the iterator.
+    /// REQUIRES: Valid()
+    #[inline]
+    pub fn key(&self) -> &[u8] {
+        unsafe {
+            let mut buf = r(&[]);
+            tirocks_sys::crocksdb_iter_key(self.ptr, &mut buf);
+            s(buf)
+        }
+    }
+
+    /// Return the value for the current entry.  The underlying storage for
+    /// the returned slice is valid only until the next modification of
+    /// the iterator.
+    /// REQUIRES: Valid()
+    #[inline]
+    pub fn value(&self) -> &[u8] {
+        unsafe {
+            let mut buf = r(&[]);
+            tirocks_sys::crocksdb_iter_value(self.ptr, &mut buf);
+            s(buf)
+        }
+    }
+
+    /// Return the sequence number for the current entry if it's available.
+    /// Return false if it's not available.
+    /// REQUIRES: Valid()
+    #[inline]
+    pub fn sequence_number(&self) -> Option<SequenceNumber> {
+        unsafe {
+            let mut seqno = 0;
+            let b = tirocks_sys::crocksdb_iter_seqno(self.ptr, &mut seqno);
+            if b {
+                Some(seqno)
+            } else {
+                None
+            }
+        }
+    }
+
+    /// If an error has occurred, return it.  Else return an ok status.
+    /// If non-blocking IO is requested and this operation cannot be
+    /// satisfied without doing some IO, then this returns Status::Incomplete().
+    #[inline]
+    pub fn check(&self) -> Result<()> {
```

Comment:
how about calling it `status` as how TiKV does currently

Reply:
For my practice, people won't use `status` directly. Instead, checking error and return early is the most common usage pattern.